src/v/cloud_storage/remote_partition.cc

```diff
@@ -348,14 +348,33 @@ void remote_partition::gc_stale_materialized_segments() {
     for (auto& st : _materialized) {
         if (
           now - st.atime > stm_max_idle_time
-          && !st.segment->download_in_progress() && st.segment.owned()) {
-            vlog(
-              _ctxlog.debug,
-              "reader for segment with base offset {} is stale",
-              st.offset_key);
-            // this will delete and unlink the object from
-            // _materialized collection
-            offsets.push_back(st.offset_key);
+          && !st.segment->download_in_progress()) {
+            if (st.segment.owned()) {
+                vlog(
+                  _ctxlog.debug,
+                  "reader for segment with base offset {} is stale",
+                  st.offset_key);
+                // this will delete and unlink the object from
+                // _materialized collection
+                offsets.push_back(st.offset_key);
+            } else {
+                vlog(
+                  _ctxlog.debug,
+                  "Materialized segment {} not stale: {} {} {} {} readers={}",
+                  st.manifest_key,
+                  now - st.atime > stm_max_idle_time,
+                  st.segment->download_in_progress(),
+                  st.segment.owned(),
+                  st.segment.use_count(),
+                  st.readers.size());
+
+                // Readers hold a reference to the segment, so for the
+                // segment.owned() check to pass, we need to clear them out.
+                while (!st.readers.empty()) {
+                    evict_reader(std::move(st.readers.front()));
+                    st.readers.pop_front();
+                }
+            }
```

Comment:
This is a good find. I think that we assumed that the readers are doing batch jobs so the readers will be removed eventually.

Also, maybe it makes sense to force eviction of such segment earlier, what do you think?

```suggestion
             if (!st.segment.owned()) {
                vlog(
                  _ctxlog.debug,
                  "Materialized segment {} not stale: {} {} {} {} readers={}",
                  st.manifest_key,
                  now - st.atime > stm_max_idle_time,
                  st.segment->download_in_progress(),
                  st.segment.owned(),
                  st.segment.use_count(),
                  st.readers.size());

                // Readers hold a reference to the segment, so for the
                // segment.owned() check to pass, we need to clear them out.
                while (!st.readers.empty()) {
                    evict_reader(std::move(st.readers.front()));
                    st.readers.pop_front();
                }
            }
            if (st.segment.owned()) {
                vlog(
                  _ctxlog.debug,
                  "reader for segment with base offset {} is stale",
                  st.offset_key);
                // this will delete and unlink the object from
                // _materialized collection
                offsets.push_back(st.offset_key);
            }
```

Reply:
Do you think it's possible that we would pass the second `st.segment.owned` check after putting the readers on the eviction queue?  It seemed to me that since there was no await/scheduling point, we would always have to wait until our next GC pass to clean up the segment